README.md
pack64
======
Pack64 is a vector encoding, with code for encoding and decoding in Python,
Ruby, Java and JavaScript.  It packs a vector into a kind-of-floating-point,
kind-of-base64 representation, requiring only 3 bytes per vector entry.

This is meant for transmitting vector data in situations where:

* Arbitrary bytes can't be transmitted
* We need to send the vector in as few bytes as possible
* Simply base64-encoding floating-point data -- at 5.33 bytes per entry --
  isn't small enough
* Some loss of precision is acceptable

Possible applications include transmitting a vector in a URL or JSON object.

If you wonder why you'd need this, the fact is that you might not.  But at
Luminoso, we send *lots* of vectors over the network.


Specifications
==============

pack64
------
Returns a string *b* of bytes, representing digits using the URL-safe base64
character set `(A-Z, a-z, 0-9, -, _)`, as follows:

* b[0] contains the power-of-two exponent, biased by 40.  That is:

    - An exponent of 0 (`A`) means to multiply all the integers that follow by
      2^-40.
    - An exponent of 30 (`e`) means to multiply the integers by 2^-10 (that is,
      divide them by 1024).
    - An exponent of 40 (`o`) means to leave the integers as is.
    - An exponent of 63 (`_`) means to multiply the integers by 2^23.

  Call this number 2^(b[0] - 40) the *increment*.

* The increment is chosen to maximize precision. To choose the increment, find
  the number "a" such that 2^a is *larger* than all the magnitudes in the
  vector.  The correct value for the increment is then 2^(a - 17).  However, if
  this gives an increment lower than 2^-40, use 2^-40 instead.  For example,
  the zero-length vector is encoded as `A`, and a length-two zero vector is
  encoded as `AAAAAAA`.

* b[1:4], b[4:7], etc. contain the values in the vector, as 18-bit,
  big-endian, twos-complement integers, which will all be multiplied by
  the increment. That is:

    - `AAA` represents 0.
    - `AAB` represents 1.
    - `AAC` represents 2.
    - `___` represents -1.
    - `__-` represents -2.
    - `f__` represents the highest possible value, (2^17 - 1).
    - `gAA` represents the lowest possible value, -(2^17).

The last value will be found in `b[3*K-2:3*K+1]`, so the length of the string
overall will be 3*K + 1.

This encoding can represent values in the range [-2^40, 2^40 - 2^23], with
granularity potentially as small as 2^-40.  It cannot represent inf or nan, and
encoders should report an error when those are encountered; you probably don't
want to transmit them in a vector anyhow.  Encoders should likewise report an
error if they encounter values outside the representable range, which may, for
speed, include -2^40 itself.

unpack64
--------
Decodes a vector that has been encoded with pack64, returning an appropriate
data structure.  In Python, this returns a NumPy vector of dtype `np.float32`;
in JavaScript, it returns a standard array of numbers.


Precision
=========

The largest entry in a pack64 vector is specified with 17 bits of precision.
This is approximately 3/4 of the precision of an IEEE 754 single-precision
float (24 bits), and corresponds to five significant (decimal) digits.

Importantly, however, every value in the vector is specified with the same
*absolute* precision.  The exponent defines 2^18 possible values for each
entry, and each entry chooses from these *same* values.  Thus entries of a
vector that are much smaller than the largest entry will lose relative
precision, possibly even being rounded to zero.  However, for transmitting
mathematical vectors, the relative precision of smaller entries generally
matters less.


License
=======
(c) 2012-2023 Luminoso Technologies.  pack64 is released as free software under
the MIT license.  See LICENSE for the terms of the MIT license (there aren't
many).


CHANGELOG.md
Python
======

### 2.0.2

* Updated Python code to work with newer NumPy

### 2.0.1

* The decoder now uses `np.frombuffer` instead of the deprecated `np.fromstring`.
* Ported tests from nosetests to pytest.

### 2.0.0

This version introduces a NumPy version dependency, and has only been tested in
Python 2.7 and 3.4+.  If these compatibility requirements pose a problem, the
previous version (1.0.3) should continue to work for the foreseeable future.

* Add changelog
* Remove functions that were not part of the intended public interface
* Remove the long-deprecated "rounded" parameter
* Add more robust input checking for unpack64(), and the option to disable it
* Significantly improve performance
* General cleanups


test.py
from pack64 import pack64, unpack64
import numpy as np

def test_random_vectors():
    for iter in xrange(900):
        magnitude = pow(2., (iter-500)/10.)
        length = (iter % 10) + 1
        vec = np.random.normal(size=(length,)) * magnitude
        if np.max(np.abs(vec)) < 2**40:
            yield round_trip_check, vec

def test_specific_vectors():
    yield round_trip_check, []
    yield round_trip_check, [0., 0., 0.]
    yield round_trip_check, [-1.]
    yield round_trip_check, [1.]
    yield round_trip_check, [1., 2.]
    yield round_trip_check, [.5, .25]

def test_encoding():
    assert pack64([]) == 'A'
    assert pack64([0.]) == 'AAAA'
    assert pack64([1.]) == 'YQAA'
    assert pack64([-1., 1.]) == 'YwAAQAA'
    assert pack64([2.**16, -1.]) == 'oQAA___'
    assert pack64([2.**16, 2**17-1]) == 'oQAAf__'
    assert pack64([2.**16, 2**17-0.2], rounded=False) == 'oQAAf__'
    assert pack64([2.**16, 2**17-0.2]) == 'pIAAQAA'
    assert pack64([2.**16, -2**17+0.2]) == 'pIAAwAA'
    assert pack64([2.**20, -1.]) == 'sQAAAAA'

def round_trip_check(vec):
    newvec = unpack64(pack64(vec, rounded=True))
    if len(vec) == 0:
        precision = 0.
        maxdiff = 0.
    else:
        precision = np.max(np.abs(vec)) * (2**-17) + 2**-40
        maxdiff = np.max(np.abs(newvec - vec))
    assert np.allclose(newvec, vec, 1e-10, precision),\
        "%s isn't close enough to %s; difference=%s, precision=%s" % (newvec,
                vec, maxdiff, precision)



python/test.py
from __future__ import unicode_literals

import numpy as np
from pack64 import pack64, unpack64
import pytest


def _check(vector, expected=None, exact=False):
    # Checks that the given vector:
    #   * Encodes successfully (and to a particular string, if specified)
    #   * Decodes successfully to exactly the same value (if specified) or to
    #     within the expected tolerance (see below for what this tolerance is)
    #   * Produces exactly the same string when the decoded value is reencoded
    # Returns the maximum absolute deviation between the given and decoded
    # vectors, and the tolerance to which it was compared.
    encoded = pack64(vector)
    if expected is not None:
        assert encoded == expected
    decoded = unpack64(encoded)
    assert pack64(decoded) == encoded

    if not len(vector):
        deviation = 0.0
    else:
        deviation = np.max(np.abs(decoded - vector))
    if exact:
        tolerance = 0.0
    else:
        # Generally pack64 guarantees a precision of 2 ** -17 times the largest
        # magnitude entry.  However, we have to adjust for two details.
        #   * The largest magnitude entry may be rounded for packing in such a
        #     way that the precision is slightly less than that guarantee.
        #   * The smallest positive number that can be packed at all is
        #     2 ** -40, so the absolute precision available for very small
        #     vectors, regardless of the size of the vector, is 2 ** -41.
        tolerance = max(np.max(np.abs(vector)) / (2.0 ** 17 - 0.5), 2.0 ** -41)
    assert deviation <= tolerance
    return deviation, tolerance


def test_specific_vectors():
    # Check small integers times a power of two, which can be encoded exactly
    _check([], expected='A', exact=True)
    _check([0.0], expected='AAAA', exact=True)
    _check([1.0], expected='YQAA', exact=True)
    _check([-1.0, 1.0], expected='YwAAQAA', exact=True)
    _check([2.0, 4.0], expected='aIAAQAA', exact=True)
    _check([0.25, 0.5], expected='XIAAQAA', exact=True)

    # Check rounding behavior
    _check([2.0 ** 17 - 1.0], expected='of__', exact=True)
    _check([2.0 ** 17], expected='pQAA', exact=True)
    _check([2.0 ** 17 - 0.6], expected='of__')  # Rounds down
    _check([2.0 ** 17 - 0.4], expected='pQAA')  # Rounds up
    _check([2.0 ** 17 - 0.5], expected='pQAA')  # Rounds up towards even
    _check([2.0 ** 17 - 1.5], expected='of_-')  # Rounds down towards even
    _check([2.0 ** -50], 'AAAA')  # Rounds to zero, does not crash

    # Demonstrate vectors at the edge of exact representability
    _check([2.0 ** 16, -1.0], expected='oQAA___', exact=True)
    _check([2.0 ** 17, -1.0], expected='pQAAAAA')  # -1.0 rounds towards zero
    _check([2.0 ** 17, -1.5], expected='pQAA___')  # -1.5 rounds towards -2.0

    # Demonstrate a vector that could be encoded exactly as 'XgAAAAB' but isn't
    _check([-1.0, 2.0 ** -17], expected='YwAAAAA')

    # Check for any intermediate truncation which would cause misencoding
    _check([2.0 ** 16 + 0.5001], expected='oQAB')
    _check([np.float32(2.0 ** 16 + 0.5001)], expected='oQAA')

    # Show that the tolerance used by the test code is as tight as possible
    # The maximum possible deviation in the absence of underflow
    deviation, tolerance = _check([2.0 ** 17 - 0.5, 1.0], expected='pQAAAAA')
    assert deviation == tolerance
    # The maximum possible deviation caused by underflow
    deviation, tolerance = _check([2.0 ** -41], expected='AAAA')
    assert deviation == tolerance


def test_input_types():
    # Both strings and bytestrings are unpackable
    assert np.all(unpack64('abcd') == unpack64(b'abcd'))

    # Anything that can be converted to a NumPy array is packable
    assert pack64([1.0, 2.0]) == 'ZIAAQAA'
    assert pack64((1.0, 2.0)) == 'ZIAAQAA'
    assert pack64(np.array([1.0, 2.0], dtype=np.float32)) == 'ZIAAQAA'
    assert pack64(np.array([1.0, 2.0], dtype=np.float64)) == 'ZIAAQAA'
    assert pack64(np.array([1.0, 2.0], dtype=int)) == 'ZIAAQAA'


def test_errors():
    # Nonfinite values are rejected
    for value in (float('inf'), float('nan')):
        with pytest.raises(ValueError):
            pack64([value])

    # Out of range values are rejected; check near the edge of the range
    with pytest.raises(OverflowError):
        pack64([(2.0 ** 17 - 0.5) * 2.0 ** 23])
    _check([(2.0 ** 17 - 0.6) * 2.0 ** 23], expected='_f__')
    with pytest.raises(OverflowError):
        # (This could actually be encoded as '_gAA'.)
        pack64([-(2.0 ** 17 - 0.5) * 2.0 ** 23])
    _check([-(2.0 ** 17 - 0.6) * 2.0 ** 23], expected='_gAB')

    # Strings with bad lengths or characters are rejected
    for string in ('', 'xx', b'xx', '\U0001f43c', 'Hey!', 'panda', 'rutabaga'):
        with pytest.raises(ValueError):
            unpack64(string)

    # Some (but not all) bad strings are accepted if error checking is disabled
    for string in ('xx', 'Hey!', 'panda'):
        unpack64(string, check=False)
    with pytest.raises(ValueError):
        unpack64('rutabaga', check=False)


def test_random_vectors():
    for magnitude in range(-45, 45):
        scale = 2.0 ** magnitude
        for length in range(1, 52, 5):
            vec = np.random.normal(scale=scale, size=(length,))
            # As demonstrated above, this is the range of encodable values
            if np.max(np.abs(vec)) < (2.0 ** 17 - 0.5) * 2.0 ** 23:
                _check(vec)


test_compatibility.py
"""
This file tests that our pack64 and unpack64 functions are compatible with the
versions in Commonsense Computing's csc_utils, which we often need to
interoperate with.

We reimplemented them from the specification to be sure that we could release
pack64/unpack64 under the MIT license.
"""

from pack64 import pack64, unpack64
from csc_utils.vector import pack64_check as reference_pack64, unpack64 as reference_unpack64
import numpy as np
import time

def test_random_vectors():
    for iter in xrange(800):
        magnitude = pow(2., (iter-500)/10.)
        length = (iter % 10) + 1
        vec = np.random.normal(size=(length,)) * magnitude
        try:
            reference_unpack64(reference_pack64(vec))
            yield encoding_check, vec
            yield decoding_check, vec
        except (OverflowError, ValueError):
            # this is a vector we can't encode in the original system
            pass

def test_specific_vectors():
    yield encoding_check, []
    yield decoding_check, []
    yield encoding_check, [0., 0., 0.]
    yield decoding_check, [0., 0., 0.]
    yield encoding_check, [-1.]
    yield decoding_check, [-1.]
    yield encoding_check, [1.]
    yield decoding_check, [1.]
    yield encoding_check, [1., 2.]
    yield decoding_check, [1., 2.]
    yield encoding_check, [.5, .25]
    yield decoding_check, [.5, .25]

def encoding_check(vec):
    a = reference_pack64(vec)
    b = pack64(vec, rounded=False)
    assert a == b,\
            '%s should have encoded to %s, got %s' % (vec, a, b)

def decoding_check(vec):
    encoded = reference_pack64(vec)
    a = reference_unpack64(encoded)
    b = unpack64(encoded)
    assert np.allclose(a, b), '%s should have decoded to %s, got %s' % (encoded, a, b)

def test_speed():
    vectors = [np.random.normal(size=(i%40+1,)) for i in xrange(40)]
    start1 = time.time()
    for vec in vectors:
        reference_unpack64(reference_pack64(vec))
    time_reference = (time.time() - start1)*1000
    start2 = time.time()
    for vec in vectors:
        unpack64(pack64(vec))
    time_ours = (time.time() - start2)*1000
    assert time_ours < time_reference,\
        "Took %4.4f ms. Time to beat: %4.4f ms." % (time_ours, time_reference)



python/pack64.py
'''
Pack64 is a vector encoding using a kind-of-floating-point, kind-of-base64
representation requiring only 3 bytes per vector entry.  This Python module
provides functions for encoding and decoding pack64 vectors.
'''

__all__ = ['pack64', 'unpack64']

import math
import numpy as np


# CHARS is a bytestring of the 64 characters in the encoding, in order.
CHARS = b'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_'

# Make lookup tables for those characters.
# DIGIT_TO_CHAR maps the numbers 0-63 to those characters.
DIGIT_TO_CHAR = np.frombuffer(CHARS, dtype=np.uint8)

# CHAR_TO_DIGIT maps byte values of those characters to the numbers 0-63.
# Other byte values are mapped to -1, and do not appear in VALID_CHARS.
CHAR_TO_DIGIT = np.full((128,), -1, dtype=int)
CHAR_TO_DIGIT[DIGIT_TO_CHAR] = np.arange(64)
VALID_CHARS = set(CHARS)

# Compute a small value that we use in determining what the exponent should
# be (more on this in a moment):
EPSILON = (2.0 ** 17 - 0.5) * 2.0 ** -40

# EPSILON is used to determine the biased exponent emitted by pack64().  The
# largest integer part that can be emitted is (2 ** 17 - 1), and hence the
# largest number that can be emitted with a biased exponent of zero is
# (2 ** 17 - 1) * (2 ** -40).  However, numbers slightly larger than this still
# round down to this when encoded.  The smallest positive number that (rounded)
# requires a biased exponent of 1 is EPSILON.  By extension, the smallest
# number that requires a biased exponent of 2 is 2 * EPSILON, and so on.

# Thus, the biased exponent that should be used can be found by applying the
# following rule to the magnitude L of the largest vector entry:
#   If L is less than...            Set the biased exponent to ...
#     EPSILON                         0
#     2 * EPSILON                     1
#     4 * EPSILON                     2
#     ...                             ...
# Or, put another way:
#   If L / EPSILON is less than...  Set the biased exponent to...
#     1 (i.e. 2 ** 0)                 0
#     2 (i.e. 2 ** 1)                 1
#     4 (i.e. 2 ** 2)                 2
#     ...                             ...
# So the biased exponent should be the smallest nonnegative integer e such that
#     L / EPSILON == m * (2 ** e)
# with m < 1.  This is exactly the computation provided by math.frexp(),
# except that we have to handle "nonnegative" ourselves.

# Because we do not check directly that the packed values fit in 18 bits, a
# rounding error in this calculation could cause a large error in the result.
# Fortunately, the values of M where we need to transition between exponents
# are all power-of-two multiples of EPSILON, which are represented exactly and
# produce exact results when divided by EPSILON and passed to frexp().

# Finally, note that this calculation ignores the availablity of 'gAA' for
# -(2 ** 17).  Using this string would allow us to encode a small number of
# vectors with greater precision, but it doesn't seem worth the effort.


def pack64(vector):
    '''
    Encode the given vector, returning a string.  Accepts any object that can
    be converted to a NumPy float array.
    '''
    if not len(vector):
        return 'A'
    vector = np.asarray(vector)

    largest_entry = np.max(np.abs(vector))
    if not np.isfinite(largest_entry):
        raise ValueError('Vector contains an invalid value.')
    biased_exponent = max(math.frexp(float(largest_entry) / EPSILON)[1], 0)
    if biased_exponent > 63:
        raise OverflowError('Vector has an entry too large to encode.')

    values = np.round(vector * 0.5 ** (biased_exponent - 40)).astype(int)
    digits = np.empty((3 * len(values) + 1,), dtype=int)
    digits[0] = biased_exponent
    digits[1::3] = values >> 12
    digits[2::3] = values >> 6
    digits[3::3] = values
    digits &= 63
    return DIGIT_TO_CHAR[digits].tobytes().decode('ascii')


def unpack64(string, check=True):
    '''
    Decode the given string, returning a NumPy array of dtype float32.
    Optionally pass check=False to disable input validation, for circumstances
    where you are sure the input is a properly packed vector.
    '''
    if isinstance(string, bytes):
        bstring = string
    else:
        bstring = string.encode('ascii')

    if check and (len(bstring) % 3 != 1 or not VALID_CHARS.issuperset(bstring)):
        raise ValueError('Cannot decode string %r' % string)
    digits = CHAR_TO_DIGIT[np.frombuffer(bstring, dtype=np.uint8)]
    values = (digits[1::3] << 12) + (digits[2::3] << 6) + digits[3::3]
    values -= (values >> 17) << 18
    return values.astype(np.float32) * 2.0 ** (digits[0] - 40)


ruby/lib/pack64.rb
# Pack64 is an encoding of vectors as strings. It packs a vector into a
# kind-of-floating-point, kind-of-base64 representation, requiring only
# 3 bytes per vector entry.
#
# See README.markdown for more details on the format.
#
# This Ruby gem provides support for Pack64 to Ruby. Strings get a method
# called .unpack64, which converts them into a Vector. Arrays and Vectors
# get .pack64 which turns them into a Pack64 string.
require 'matrix'

module Pack64
  CHARS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_'
  SIGN_BIT = 131072
  ROUND_MARGIN = SIGN_BIT / (SIGN_BIT - 0.5)
  INFINITY = +1.0/0.0
  NAN = 0.0/0.0

  $chars_to_indices = {}

  CHARS.split('').each_with_index do |char, i|
    $chars_to_indices[char] = i
  end

  # Returns a compact string encoding that approximates an N-dimensional vector.
  # Can take in either a Vector or a standard array.
  def Pack64.pack64(vector)
    array = vector.to_a
    return 'A' if array.length == 0
    absolute = array.map { |x| x.abs }
    highest = absolute.max * ROUND_MARGIN
    if highest.infinite? or highest.nan?
      fail FloatDomainError.new('Vector contains an invalid value')
    end

    if highest == 0
      lowest_unused_power = -40
    else
      lowest_unused_power = Pack64.log2(highest).floor + 1
      if lowest_unused_power > 40
        fail RangeError.new('Overflow')
      end
    end

    exponent = [lowest_unused_power - 17, -40].max
    increment = 2**exponent
    first = exponent + 40
    newarray = array.map { |x| (x / (increment.to_f)).round }

    chunks = [CHARS[first]] + newarray.map { |x| Pack64.twos_complement_encode(x) }
    return chunks.join('')
  end

  # Decode the given string (encoded using pack64) into a Vector.
  def Pack64.unpack64(string)
    increment = 2**($chars_to_indices[string[0]] - 40).to_f
    rest = string[1..-1]
    numbers = []
    0.upto(rest.length / 3 - 1) do |index|
      chunk = rest[(index * 3)...(index * 3 + 3)]
      val = Pack64.twos_complement_decode(chunk)
      numbers.push(val * increment)
    end
    return Vector.elements(numbers)
  end

private

  def Pack64.log2(num)
    return Math.log(num) / Math.log(2)
  end

  # Given a number, return a three-character string representing it
  # (rounded to the nearest int), as 18-bit two's complement in base64.
  def Pack64.twos_complement_encode(number)
    number = number.round
    if number < -SIGN_BIT || number >= SIGN_BIT
      fail RangeError.new("Integer out of range: #{number}")
    end

    number += SIGN_BIT * 2 if number < 0

    first = number / 4096
    without_firstval = number - 4096 * first
    second = without_firstval / 64
    third = without_firstval - 64 * second
    return CHARS[first] + CHARS[second] + CHARS[third]
  end

  # Given a three-character string (encoded from twosComplementEncode),
  # return the integer it represents.
  def Pack64.twos_complement_decode(string)
    number = 4096 * $chars_to_indices[string[0]] \
             + 64 * $chars_to_indices[string[1]] \
             +      $chars_to_indices[string[2]]
    if number >= SIGN_BIT
      number -= SIGN_BIT * 2
    end
    number
  end

end

# Now include pack64 and unpack64 as methods of Strings, Vectors, and Arrays.
class String
  def unpack64
    Pack64.unpack64(self)
  end
end

class Vector
  def pack64
    Pack64.pack64(self)
  end
end

class Array
  def pack64
    Pack64.pack64(self)
  end
end



java/src/main/java/com/luminoso/pack64/Pack64.java
package com.luminoso.pack64;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * pack64.pack: encodes a vector into a pack64'd string.
 * pack64.unpack: decodes a pack64'd string into a vector.
 *
 * This library is for decoding a packed vector format, defined originally
 * in the Python package `pack64`.
 * 
 * The format uses URL-safe base64 to encode an exponent followed by several
 * 18-bit signed integers.
 *
 * @copyright
 * (c) Copyright 2019 Luminoso Technologies, Inc
 * 
 * @license
 * MIT License - See LICENSE file
 * 
 */

public class Pack64 {

    // 2^17 is the number that makes an 18-bit signed integer go negative.
    static int SIGN_BIT = 131072;
    static Double ROUND_MARGIN = SIGN_BIT / (SIGN_BIT - 0.5);

    static char[] alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_".toCharArray();
    private static final Map<Character,Integer> alphabet_map;
    static {
        Map<Character, Integer> aMap = new HashMap<Character,Integer>();
        int i;
        for (i = 0; i < alphabet.length; i++) {
            aMap.put(alphabet[i],i);
        }
        alphabet_map = Collections.unmodifiableMap(aMap);
    }

    /**
     * pack
     * @param double[] an arraby of doubles that make up the vector
     *
     * @return Returns an url save base64 string that represents the vector
     */
    public static String pack(double[] vector) {
        if (vector.length == 0)
            return "A";

        // Calculate the smallest power of 2 we *don't* need to represent.
        // The exponent we want will be 17 lower than that.
        double max = 0;
        int i;
        for (i = 0; i < vector.length; i++) {
            double v = Math.abs(vector[i]) * ROUND_MARGIN;
            if (v > max) {
                max = v;
            }
        }

        int upperBound;
        if (max == 0)
            upperBound = -40;
        else
            upperBound = (int) Math.floor(1 + Math.log(max) / Math.log(2));

        int exponent = upperBound - 17;
        if (exponent > 23) {
            // Overflow. Return the flag vector for "almost infinity".
            return String.valueOf('-');
        }

        if (exponent < -40) {
            // Underflow. Lose some precision. Or maybe all of it.
            exponent = -40;
        }

        double power = Math.pow(2, exponent);

        StringBuilder res = new StringBuilder();
        res.append(alphabet[exponent + 40]);

        for (i = 0; i < vector.length; i++) {
            int num = (int) Math.round((vector[i] / power));
            if (num < 0) {
                num += SIGN_BIT*2;
            }
            // Do the signed arithmetic to represent an 18-bit integer.
            res.append(alphabet[(num % (1 << 18)) >> 12]);
            res.append(alphabet[(num % (1 << 12)) >> 6]);
            res.append(alphabet[num % (1 << 6)]);
        }

        return res.toString();
    }

    /**
     * unpack
     *
     * @param String a base64 representation of a vector created with pack
     *
     * @return an array of double values representing the vector
     */
    public static double[] unpack(String stVect) {
        ArrayList<Integer> hexes = new ArrayList<Integer>();
        int i;
        for (i = 0; i < stVect.length(); i++) {
            hexes.add(alphabet_map.get(stVect.charAt(i)));
        }

        int K = (hexes.size() - 1) / 3;
        double[] vector = new double[K];
        double unit = Math.pow(2, hexes.get(0) - 40);
        for (i = 0; i < K; i++) {
            int base = i * 3;
            int integer = 4096 * hexes.get(base + 1) + 64 * hexes.get(base + 2) + hexes.get(base + 3);
            if (integer >= SIGN_BIT) {
                integer -= SIGN_BIT * 2;
            }
            vector[i] = (integer * unit);
        }

        return vector;
    }
}

javascript/pack64.js
/*
 * pack64.pack: encodes a vector into a pack64'd string.
 * pack64.unpack: decodes a pack64'd string into a vector.
 *
 * This library is for decoding a packed vector format, defined in the Python
 * package `pack64`.
 * 
 * The format uses URL-safe base64 to encode an exponent followed by several
 * 18-bit signed integers.
 */
(function (name, global, definition) {
    if (typeof module !== 'undefined') {
        module.exports = definition();
    } else if (typeof define !== 'undefined' && typeof define.amd === 'object') {
        define(definition);
    } else {
        global[name] = definition();
    }
})('pack64', this, function () {
    "use strict";

    // 2^17 is the number that makes an 18-bit signed integer go negative.
    var SIGN_BIT = 131072;
    var ROUND_MARGIN = SIGN_BIT / (SIGN_BIT - 0.5);

    var alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";
    var alphabet_map = {};
    for (var i = 0; i < 64; i++) {
        alphabet_map[alphabet.charAt(i)] = i;
    }

    function pack(vec) {
        // Calculate the smallest power of 2 we *don't* need to represent.
        // The exponent we want will be 17 lower than that.
        var max = 0, i;
        for (i = 0; i < vec.length; i++) {
            var v = Math.abs(vec[i]) * ROUND_MARGIN;
            if (v > max) {
                max = v;
            }
        }

        var upperBound = Math.floor(1 + Math.log(max) / Math.log(2));
        var exponent = upperBound - 17;
        if (exponent > 23) {
            // Overflow. Return the flag vector for "almost infinity".
            return '-';
        }

        if (exponent < -40) {
            // Underflow. Lose some precision. Or maybe all of it.
            exponent = -40;
        }

        var power = Math.pow(2, exponent);
        var res = [alphabet[exponent + 40]];
        for (i = 0; i < vec.length; i++) {
            var num = Math.round(vec[i] / power);
            if (num < 0) {
                num += SIGN_BIT*2;
            }
            // Do the signed arithmetic to represent an 18-bit integer.
            res.push(alphabet[(num % (1 << 18)) >> 12]);
            res.push(alphabet[(num % (1 << 12)) >> 6]);
            res.push(alphabet[num % (1 << 6)]);
        }

        return res.join('');
    }

    function unpack(str) {
        var hexes = [];
        for (var i = 0; i < str.length; i++) {
            hexes[i] = alphabet_map[str.charAt(i)];
        }

        var K = (hexes.length - 1) / 3;
        var vector = new Float32Array(Math.ceil(K));
        var unit = Math.pow(2, hexes[0] - 40);
        for (var i = 0; i < K; i++) {
            var base = i * 3
            var integer = 4096*hexes[base + 1] + 64*hexes[base + 2] + hexes[base + 3];
            if (integer >= SIGN_BIT) {
                integer -= SIGN_BIT * 2;
            }
            vector[i] = integer * unit;
        }

        return vector;
    }

    function test() {

        function assertPackedEquals(v, str) {
            console.assert(pack(v) == str, "pack([" + v + "]) == " + str)
        }

        assertPackedEquals([ ], 'A')
        assertPackedEquals([0], 'AAAA')
        assertPackedEquals([1],'YQAA')
        assertPackedEquals([-1, 1], 'YwAAQAA')
        assertPackedEquals([Math.pow(2, 16), -1], 'oQAA___')
        assertPackedEquals([Math.pow(2, 16),Math.pow(2, 17) - 1], 'oQAAf__')
        assertPackedEquals([Math.pow(2, 16),Math.pow(2, 17) - 0.2], 'pIAAQAA')
        assertPackedEquals([Math.pow(2, 16), Math.pow(-2, 17) + 0.2], 'pIAAwAA')
        assertPackedEquals([Math.pow(2, 20), -1], 'sQAAAAA')

        function assertRoundtripIsEqual(v) {
            var roundtripped = unpack(pack(v))

            for (var i = 0; i < roundtripped.length; i++) {
                console.assert(roundtripped[i] == v[i],
                    "pack(unpack([" + v + "])) == [" + v + "]")
            }

            console.assert(roundtripped.length == v.length,
                "The length of the vector stays the same after a unpack(pack(v)) cycle. ([" + v + "] != [" + roundtripped + "]) ")
        }
        assertRoundtripIsEqual([1, 2, 3])
        assertRoundtripIsEqual([])
        assertRoundtripIsEqual([0, 0, 0])
        assertRoundtripIsEqual([-1])
        assertRoundtripIsEqual([1])
        assertRoundtripIsEqual([1, 2])
        assertRoundtripIsEqual([.5, .25])

    }

    return { pack: pack, unpack: unpack, test: test };
});
